### ISSUE TITLE
Node resolution server capabilities and more

### ISSUE THREAD

**Issue by daKmoR** [microsoft/vscode-livepreview#170]
hey,

I did not know about this project and I have been playing around with combining `@web/dev-server` with a vscode extension.

With that the server can handle reloading and especially the server can handle node resolution of javascript files.

e.g. I can write

```html
<script type="module">
  import { Foo } from 'my-pkg';
  console.log(Foo);
</script>
```

and it will import it from `path/to/node_modules/my-pkg/index.js` and the path can even be outside of the server root. (a special feature of [@web/dev-server](https://modern-web.dev/docs/dev-server/overview/))

I took a quick look at the code and the server in this extension seems to be pretty tailor maid so I guess getting an option to switch out the server will not be possible? 🤔

Here is a preview of my current POC

it has some different features like
- supporting node resolution
- switching files changes the preview to the same

one "downside" is that you need to save to get the changes... I'm not sure if I would prefer this over "instant" changes 🤔

I did a tweet today about it
https://twitter.com/daKmoR/status/1457755974815690755?s=20

here is the video from that tweet

https://user-images.githubusercontent.com/24378/140801466-c0b6152b-1ecc-43a8-bffb-da8536bbc8c2.mp4

I'm not really sure how to proceed?

Here are some ideas:
- Should I keep a separate extension? (and definitely get some inspirations from here)
- Should I drop it?
- I was also thinking of enabling that a user can provide a web-dev-server.config.js file - no idea yet how that would work?
- Would you consider supporting node resolution?
- Would you consider (or accept a contribution) to switch out the server with `@web/dev-server`? 
- ...

I am completely open for any suggestions/feedback... 🤗

**Comment by lszomoru** [comment]
Adding @digitarald for visibility.

**Comment by digitarald** [comment]
First thought I had would be that users could configure a preprocessor (either CLI or as extension contribution) for their files, but what command would make this snippet work:

```
<script type="module">
  import { Foo } from 'my-pkg';
  console.log(Foo);
</script>
```

It seems you want a polyfill for supporting modules in the browser, but you also seem to expect that it is resolved via node.js? Thanks for clarifying, @daKmoR.

@lszomoru what do you think about having extension contributions for handling the probably growing list of file pre-processing? Is there an issue related to that

**Comment by daKmoR** [comment]
To clarify it's not solely for the node resolution... it's about that most dev servers will inevitably at some point require pretty heavy customization if you wanna use them at "bigger" projects 😅

for example, supporting
- SPA routing
- [Proxying requests to an API](https://modern-web.dev/docs/dev-server/middleware/#proxying-requests)
- [Range Requests](https://modern-web.dev/docs/dev-server/middleware/#enabling-range-requests)
- Hot Module Replacement
- [Reusing Rollup Plugins](https://modern-web.dev/docs/dev-server/plugins/rollup/) or any other single file transform system
- Typescript
- Esbuild
- [build your own plugin](https://modern-web.dev/docs/dev-server/writing-plugins/overview/)
- access to websockets
- ...

In short I wanna avoid that I have my local dev sever and the vs code dev server with vastly different capabilities... hence the idea to be it "the same"

I can totally understand if that is not "in scope" and the focus is more on editing HTML... if so then a separate extension might make sense even though it would probably be pretty much the same except for the server 🙈

**Comment by justinfagnani** [comment]
I came here to look for or file this issue, and I want to chime in that I think module resolution is by far the most important feature for any pre-processing. This is because in even in a plain JavaScript or TypeScript project with no other build system or framework, both the first party and third party modules will use bare specifiers in imports.

IMO, the other things that @daKmoR lists are nice, but very optional and could be brought by a dev server like `@web/dev-server`, Vite, or Turbopack.

@digitarald 
> It seems you want a polyfill for supporting modules in the browser

Browsers have supported modules for a long time now, they just don't support bare module specifiers (at least without import maps).

**Comment by andreamah** [comment]
Currently, Live Preview focuses on just previewing simple html projects. 
If you want to create your own custom server with preprocessing, you can feel free to use Simple Browser in vscode by running `Simple Browser: Show`. You can access your own hosted content via that window if you want an embedded experience.

**Comment by justinfagnani** [comment]
@andreamah I think the idea here is that any simple HTML projects that include the most simple standard JS modules will need import specifier resolution. This is true for very plain JS - no frameworks, no compilers.

Is the recommendation here to use Import Maps? Do they work in live preview?

**Comment by ltlian** [comment]
My use case for Live Preview is (as mentioned in this thread) for very simple pages that don't rely on a build step. Sometimes I still need to reference packages without wanting to involve node or npm.

To answer the last comment, import maps work perfectly fine for this scenario and just dropping the example for future adventurers. Here's what I have working on my end:

```html
  <script type="importmap">
    {
      "imports": {
        "some-lib": "./node_modules/some-lib/some-lib.module.js"
      }
    }
  </script>
```

When deploying the page, I replace the value with `"https://asdf/some-lib.module.js"`.
